Reconstruction/RecTracker/src/components/RecTrackAlg.cpp

```diff
@@ -0,0 +1,118 @@
+
+#include "DetInterface/IGeoSvc.h"
+
+#include "datamodel/PositionedTrackHitCollection.h"
+#include "datamodel/TrackHitCollection.h"
+#include "datamodel/TrackStateCollection.h"
+#include "datamodel/TrackCollection.h"
+
+#include "DD4hep/Detector.h"
+#include "DD4hep/Volumes.h"
+#include "DDRec/API/IDDecoder.h"
+#include "DDSegmentation/BitField64.h"
+#include "DDSegmentation/CartesianGridXZ.h"
+
+#include <cmath>
+#include <random>
+
+#include "RecTrackAlg.h"
+#include "RecInterface/ITrackSeedingTool.h"
+#include "RecTracker/TrackingUtils.h"
+
+#include "tricktrack/RiemannFit.h"
+
+DECLARE_ALGORITHM_FACTORY(RecTrackAlg)
+
+RecTrackAlg::RecTrackAlg(const std::string& name, ISvcLocator* svcLoc)
+    : GaudiAlgorithm(name, svcLoc) {
+
+  declareProperty("positionedTrackHits", m_positionedTrackHits, "Tracker hits (Input)");
+  declareProperty("Tracks", m_tracks, "Tracks (Output)");
+  declareProperty("TrackStates", m_trackStates, "TrackStates (Output)");
+  declareProperty("TrackSeedingTool", m_trackSeedingTool);
+}
+
+StatusCode RecTrackAlg::initialize() {
+  info() << "initialize" << endmsg;
+
+  return StatusCode::SUCCESS;
+}
+
+StatusCode RecTrackAlg::execute() {
+
+  // get hits from event store
+  const fcc::PositionedTrackHitCollection* hits = m_positionedTrackHits.get();
+  fcc::TrackCollection* tracks = m_tracks.createAndPut();
+  fcc::TrackStateCollection* trackStates = m_trackStates.createAndPut();
+  debug() << "hit collection size: " << hits->size() << endmsg;
+  auto seedmap = m_trackSeedingTool->findSeeds(hits);
```

Comment:
Since you are using a `multimap` structure in `seedmap`, maybe using its method `equal_range` would make the code easier to read on the double loop part. A quick example quick be:

```cpp
// get hits from event store
const fcc::PositionedTrackHitCollection* hits = m_positionedTrackHits.get();
fcc::TrackCollection* tracks = m_tracks.createAndPut();
fcc::TrackStateCollection* trackStates = m_trackStates.createAndPut();
debug() << "hit collection size: " << hits->size() << endmsg;
auto seedmap = m_trackSeedingTool->findSeeds(hits);

constexpr unsigned int nhits = 8;

// Iterate through multimap of <tracks-hits> elements
// range.first : first multimap element containing a range of hits with the same trackID
// range.second : last multimap element of the range
for(auto i = seedmap.begin(); i != seedmap.end(); i = range.second){

  // Get the range of the current TrackID
  range = seedmap.equal_range(i->first);

  // Go through the range
  for(auto d = range.first; d != range.second; ++d){

  }

}
```

Although your loop do pretty much the same, it may be less error-prone than handling both iterators.